datafusion/datasource-parquet/src/sort.rs

```diff
@@ -31,22 +31,34 @@ use std::collections::HashMap;
 /// 3. Reconstructs the row selection for the new order
 ///
 /// # Arguments
-/// * `row_selection` - Original row selection
+/// * `row_selection` - Original row selection (only covers row groups that are scanned)
 /// * `parquet_metadata` - Metadata containing row group information
+/// * `row_groups_to_scan` - Indexes of row groups that will be scanned (in original order)
 ///
 /// # Returns
 /// A new `RowSelection` adjusted for reversed row group order
+///
+/// # Important Notes
+/// The input `row_selection` only covers the row groups specified in `row_groups_to_scan`.
+/// Row groups that are skipped (not in `row_groups_to_scan`) are not represented in the
+/// `row_selection` at all. This function needs `row_groups_to_scan` to correctly map
+/// the selection back to the original row groups.
 pub fn reverse_row_selection(
     row_selection: &RowSelection,
     parquet_metadata: &ParquetMetaData,
+    row_groups_to_scan: &[usize],
 ) -> Result<RowSelection> {
     let rg_metadata = parquet_metadata.row_groups();
 
-    // Build a mapping of row group index to its row range in the file
+    // Build a mapping of row group index to its row range, but ONLY for
+    // the row groups that are actually being scanned.
+    // The row numbers in this mapping are relative to the scanned row groups,
+    // not the entire file.
     let mut rg_row_ranges: Vec<(usize, usize, usize)> =
-        Vec::with_capacity(rg_metadata.len());
+        Vec::with_capacity(row_groups_to_scan.len());
     let mut current_row = 0;
```

Comment:
Is the tracking for `current_row` still correct? We used to scan every row so I think it was correct, but now we if we skip a row group won't `current_row` be incorrect globally?